### PR TITLE
Access the relation databag if unit is leader

### PR DESCRIPTION
**Pull request by gatici**
# Description

This aims to fix https://github.com/canonical/tls-certificates-interface/issues/58 and  https://github.com/canonical/self-signed-certificates-operator/issues/24. Within this change, non-leader does not try to access app relation databag and Permission denied error does not appear.

## Checklist

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.
- [x] I have added tests that validate the behaviour of the software.
- [x] I validated that new and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have bumped the version of any required library.

### CODE REVIEW

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 1]
```diff
@@ -897,6 +897,22 @@ def __init__(self, charm: CharmBase, relationship_name: str):
         self.charm = charm
         self.relationship_name = relationship_name
 
+    def _load_relation_data(self, raw_relation_data: dict) -> dict:
```
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:900]: There might be cases when `_load_relation_data` should work for non-leader-units.
For example a non-leader unit trying to read the unit databag of the same application or reading from the remote application's databags.
Please refer to [this diagram ](https://discourse.charmhub.io/uploads/default/original/2X/c/cc50a23c3d5ce4458dd0ca81f3d3ff1768ad0607.png) from the [docs](https://juju.is/docs/sdk/integration).

I would suggest adding guards where we try to call this function with `some_relation.data[self.charm.app]`

Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:900]: Since it is a private method, I am not sure we need to support those cases, as there does not seem to be any of those cases in this library.

Do you have an example in mind?

Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:900]: Yes I do.

there is this line in the `request_certificate` of the requirer
`relation_data = _load_relation_data(relation.data[self.model.unit])`
I don't see why `request_certificate`  would be only done by leader units.

But why should `_load_relation_data` know about the unit and if it's leader or not? in my opinion we should guard the calls of `_load_relation_data`.

Comment (gatici) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:900]: @saltiyazan Thanks for indication the problem here. I agree, we need to add guard only  if we call some_relation.data[self.charm.app]. 
The PR is updated accordingly.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 2]
```diff
@@ -897,6 +897,22 @@ def __init__(self, charm: CharmBase, relationship_name: str):
...
+        Args:
+            raw_relation_data (dict): Relation data from the databag
+
+        Returns:
+            dict: Relation data in dict
+        """
+        # If unit is not leader, it does not try to reach relation data.
+        if not self.model.unit.is_leader():
+            return {}
+        return _load_relation_data(raw_relation_data)
```
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:914]: is it calling itself? or do you mean to have a different function here?

Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:914]: oh, never mind, I see it's calling the global function.

Comment (gatici) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:914]: _load_relation_data is the original function here. Above is the method under TLSCertificatesProvidesV2 that calls the function.

**Comment by saltiyazan** [comment]
Are unit test cases aligned with this change?

**Comment by gatici** [comment]
> Are unit test cases aligned with this change?

Tests are added.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py [hunk 3]
```diff
@@ -364,6 +489,62 @@ def test_given_some_certificates_in_relation_data_when_set_relation_certificate_
...
+        expected_relation_data = {
+            "certificates": [
+                {
+                    "certificate": initial_certificate,
+                    "certificate_signing_request": initial_certificate_signing_request,
+                    "ca": initial_ca,
+                    "chain": initial_chain,
+                },
+            ]
+        }
```
Comment (ghislainbourgeois) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:540]: This would make the test a bit more maintainable.
```suggestion
        expected_relation_data = {
            "certificates": json.loads(key_values["certificates"])
        }
```

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py [hunk 4]
```diff
@@ -414,6 +595,57 @@ def test_given_identical_csr_in_relation_data_when_set_relation_certificate_then
...
+        expected_relation_data = {
+            "certificates": [
+                {
+                    "certificate_signing_request": initial_certificate_signing_request,
+                    "certificate": initial_certificate,
+                    "ca": initial_ca,
+                    "chain": initial_chain,
+                },
+            ]
+        }
```
Comment (ghislainbourgeois) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:641]:
```suggestion
        expected_relation_data = {
            "certificates": json.loads(key_Values["certificates"])
        }
```

**Review by ghislainbourgeois — APPROVED** [review]
I have some small suggestions, but otherwise LGTM.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py (hunk 3 above)
Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py]: Done

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py (hunk 4 above)
Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py]: Done

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py [hunk 5]
```diff
@@ -185,6 +185,92 @@ def test_given_certificate_for_csr_already_in_relation_data_when_on_relation_cha
...
+        requirer_unit_data = {
+            "certificate_signing_requests": json.dumps(
+                [
+                    {
+                        "certificate_signing_request": csr,
+                    }
+                ]
+            )
+        }
+        self.harness.set_leader(is_leader=False)
```
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:225]: Why are we calling `set_leader` here?

Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:225]: We are testing unit is not leader when relation changed. This is why we are setting leader False.

Comment (saltiyazan) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:225]: But it was leader at the beginning of the test, no?

Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:225]: If we check the test definition CSR is already in the relation data. To require a leader unit to  prepare the conditions that this test needs. Then we are testing a relation changed situation with a non leader unit.

Comment (saltiyazan) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:225]: Yeah, I get it now, but in that case we are testing that we request the creation of certificate for a csr that already has a certificate, no?
Is that an expected behaviour? This is a change compared to what the library was doing before.
I think the root of this is that in the `on_relation_changed` in the provider we require the unit to be leader to get `provider_relation_data` but we don't have that requirement to get the `requirer_relation_data`
I think in this case if a unit is not leader it should not do anything.
If we use the `self._load_relation_data` in both cases we will get to that, but I again find it a bit hard to follow, I'd suggest using guards in place where needed and keeping one `_load_relation_data` or using `self._load_relation_data` only in the provider.

Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:225]: Yes, we are testing that we request the creation of certificate for a crs that already has a certificate. If the relation changed and provider unit is not leader, if can not get the certificate which is already there as we put a guard for non-leader unit. You can see the there is no guard in the requirer  part and we are using self._load_relation_data  only in the provider side if  we try to app data bag. As we can not that certificate because of guard, code process it like it does not exist.  This causes to trigger another certificate creation request.  You think this part is problematic, right ? A non leader unit should not cause to any action.

Comment (saltiyazan) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:225]: I think the problematic part is in the provider in the `_on_relation_changed`. 
The logic there is to filter CSRs that already got their certificates, now we don't do that if the provider unit is not leader, we just issue a new certificate for them.
That part needs a revisit. In my opinion it should be like it was before, no certificate should be issued to a CSR with a certificate (can be done by using guards in provider `_on_relation_changed`).
Let me know what you think and if you have a different suggestion.

Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:251]: This is very good catch. I fixed it in the latest patch. The code was comparing certificates_signing_requests in provider_csrs and requirer_csrs. However provider_csrs  was not reflecting the truth so this situation happened. I fixed this problem adding another check on_relation_changed_method before processing to those step. Now a non-leader unit does nothing.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 6]
```diff
@@ -1111,9 +1127,12 @@ def _on_relation_changed(self, event: RelationChangedEvent) -> None:
         Returns:
             None
         """
-        assert event.unit is not None
+        if event.unit is None:
```
[the remark this answers is not in the document]
Comment (gatici) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1130]: Error logging is added.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py [hunk 7]
```diff
@@ -185,6 +211,92 @@ def test_given_certificate_for_csr_already_in_relation_data_when_on_relation_cha
 
         patch_certificate_creation_request.assert_not_called()
 
+    @patch(
+        f"{LIB_DIR}.CertificatesProviderCharmEvents.certificate_creation_request",
+        new_callable=PropertyMock,
+    )
+    def test_given_certificate_for_csr_already_in_relation_data_when_on_relation_changed_and_unit_not_leader_then_certificate_creation_request_is_not_emitted(  # noqa: E501
```
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:218]: Should we keep this test now that the `is_leader` guard is added at the beginning of `on_config_changed`? I mean the cert creation request is not going to be emitted independently from wether the csr is in the databag or not.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py [hunk 8]
```diff
@@ -414,6 +612,48 @@ def test_given_identical_csr_in_relation_data_when_set_relation_certificate_then
         loaded_relation_data = _load_relation_data(dict(provider_relation_data))
         self.assertEqual(expected_relation_data, loaded_relation_data)
 
+    def test_given_identical_csr_in_relation_data_when_set_relation_certificate_and_unit_not_leader_then_certificate_is_not_replaced_in_relation_data(  # noqa: E501
```
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:615]: Like above, cases look independent. is the cert not replaced because the csr is identical or because unit is not leader?

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py [hunk 9]
```diff
@@ -848,6 +1147,65 @@ def test_given_multiple_certificate_in_relation_data_when_get_issued_certificate
         certificates = self.harness.charm.certificates.get_issued_certificates()
         self.assertEqual(certificates, expected_certificates)
 
+    def test_given_multiple_certificate_in_relation_data_when_get_issued_certificates_and_unit_not_leader_then_returned_dicts_has_empty_certificates_list(
```
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:1150]:
```suggestion
    def test_given_multiple_certificate_in_relation_data_when_get_issued_certificates_and_unit_not_leader_then_returned_dict_has_empty_certificates_list(
```

But as above, the list will be empty independent from the number of certs.

**Review by saltiyazan — APPROVED** [review]
Other than nitpicking on test cases it looks good to me.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py (hunk 7 above)
Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:218]: Here, certificate request is not emitted because unit is not leader. csr is in relation databag or not does not have any effect.This test is not beneficial anymore.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py (hunk 8 above)
Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py:615]: Likewise, I would like to remove this test, unit is not leader is main reason independent from other conditions.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py (hunk 9 above)
Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py]: This is removed.

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py (hunk 8 above)
Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py]: Done

File: tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py (hunk 7 above)
Comment (gatici) [reply tests/unit/charms/tls_certificates_interface/v2/test_tls_certificates_v2_provides.py]: Done

**Review by saltiyazan — APPROVED** [review]

**Review by Gmerold — APPROVED** [review]
